magma/passes/clock.py

```diff
@@ -1,10 +1,201 @@
+    try:
+        first_clk = next(first_clks)
+    except StopIteration:  # early exit to avoid traversing other elements
+        return None
+    yield first_clk
+    yield from first_clks
+    for elem in value[1:]:
+        yield from get_output_clocks_in_value(elem, clock_type)
+
+
+def get_output_clocks_in_value(value: Type, clock_type: Kind) -> Iterable[Type]:
+    """
+    Gets all output values of type @clock_type recursively contained in @value.
+    """
+    # Since we are looking for default drivers, only need to consider outputs.
+    if value.is_input():
+        return None
+    if isinstance(value, clock_type):
+        yield value
+    if isinstance(value, Tuple):
+        for elem in value:
+            yield from get_output_clocks_in_value(elem, clock_type)
+    if isinstance(value, Array):
+        yield from _get_output_clocks_in_array(value, clock_type)
+
+
+def get_output_clocks_in_defn(
+        defn: DefineCircuitKind, clock_type: Kind) -> Iterable[Type]:
+    """
+    Gets all output values of type @clock_type in circuit @defn. This includes
+    "input" ports of @defn and output ports of any instances contained in @defn.
+    """
+    for port in defn.interface.ports.values():
+        yield from get_output_clocks_in_value(port, clock_type)
+    for inst in defn.instances:
+        if isinstance(type(inst), Wire):
+            # We want to skip instances of Wire(T), because they are basically
+            # buffers and therefore their outputs are not legitimate
+            # drivers. Not doing so might result in spurrious "multiple clocks".
+            continue
+        for port in inst.interface.ports.values():
+            yield from get_output_clocks_in_value(port, clock_type)
+
+
+def get_all_output_clocks_in_defn(
+        defn: DefineCircuitKind) -> Mapping[Kind, Type]:
+    """
+    Returns a map from each clock type to either the singleton output value of
+    that type in @defn, or None (if there is none, or there are multiple).
+    """
+    default_clocks = {}
+    for clock_type in ClockTypes:
+        clocks = get_output_clocks_in_defn(defn, clock_type)
+        try:
+            clock = only(clocks)
+        except IterableException:
+            clock = None
+        default_clocks[clock_type] = clock
+    return default_clocks
+
+
+def _get_undriven_clocks_in_array(
+        value: Array, clock_type: Kind) -> Iterable[Type]:
+    """
+    Returns all undriven values of type @clock_type contained in array @value.
+    """
+    # First check if the first element has any undriven.
+    it = get_undriven_clocks_in_value(value[0], clock_type)
+    try:
+        first = next(it)
+    except StopIteration:  # early exit to avoid traversing other elements
+        return
+    yield first
+    yield from it
+    # NOTE(leonardt): Since magma arrays don't support zero-length values, we
+    # have to check the size before slicing.
+    if len(value) <= 1:
+        return
```

Comment:
Hmm, minor comment but if it's to be believed, then maybe we should change it to `len(value) == 1` and put an assertion that `len(value) >= 1` (or update the comment if it's possible to have a length 0 array somewhere.

Reply:
I think this comment is more about that if you do `value[1:]` in python it would just return an empty list, but with a magma array it will result in an error regarding 0-length arrays. Therefore we have to do an early exit check rather than just having a zero-length array be returned.